eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/NodeRelay.scala

```diff
@@ -223,102 +246,121 @@ class NodeRelay private(nodeParams: NodeParams,
+        // resolve that to a nodeId in order to reach that introduction node and use the blinded path.
+        // If we are the introduction node ourselves, we'll also need to decrypt the onion and identify the next node.
         context.spawnAnonymous(BlindedPathsResolver(nodeParams, paymentHash, router, register)) ! Resolve(context.messageAdapter[Seq[ResolvedPath]](WrappedResolvedPaths), payloadOut.outgoingBlindedPaths)
-        waitForResolvedPaths(upstream, payloadOut, paymentCfg, routeParams)
+        Behaviors.receiveMessagePartial {
+          rejectExtraHtlcPartialFunction orElse {
+            case WrappedResolvedPaths(resolved) if resolved.isEmpty =>
+              context.log.warn("rejecting trampoline payment to blinded paths: no usable blinded path")
+              rejectPayment(upstream, Some(UnknownNextPeer()))
+              stopping()
+            case WrappedResolvedPaths(resolved) =>
+              // We don't have access to the invoice: we use the only node_id that somewhat makes sense for the recipient.
+              val blindedNodeId = resolved.head.route.blindedNodeIds.last
+              val recipient = BlindedRecipient.fromPaths(blindedNodeId, Features(payloadOut.invoiceFeatures).invoiceFeatures(), payloadOut.amountToForward, payloadOut.outgoingCltv, resolved, Set.empty)
+              context.log.debug("forwarding payment to blinded recipient {}", recipient.nodeId)
+              wakeUpIfNeeded(upstream, recipient, nextPayload, nextPacket_opt)
+          }
+        }
     }
   }
 
   /**
-   * Blinded paths in Bolt 12 invoices may encode the introduction node with an scid and a direction: we need to resolve
-   * that to a nodeId in order to reach that introduction node and use the blinded path.
+   * The next node may be a mobile wallet directly connected to us: in that case, we'll need to wake them up before
+   * relaying the payment.
    */
-  private def waitForResolvedPaths(upstream: Upstream.Trampoline,
-                                   payloadOut: IntermediatePayload.NodeRelay.ToBlindedPaths,
-                                   paymentCfg: SendPaymentConfig,
-                                   routeParams: RouteParams): Behavior[Command] =
+  private def wakeUpIfNeeded(upstream: Upstream.Trampoline, recipient: Recipient, nextPayload: IntermediatePayload.NodeRelay, nextPacket_opt: Option[OnionRoutingPacket]): Behavior[Command] = {
+    if (isAsyncPayment(nodeParams, nextPayload)) {
+      waitForTrigger(upstream, recipient, nextPayload, nextPacket_opt)
+    } else if (shouldWakeUpNextNode(nodeParams)) {
+      waitForPeerReady(upstream, recipient, nextPayload, nextPacket_opt)
+    } else {
+      relay(upstream, recipient, nextPayload, nextPacket_opt)
+    }
+  }
+
+  /**
+   * The next node is the payment recipient. They are directly connected to us and may be offline, but supports async
+   * payments. We wait for them to be online and signal that they're ready to receive the payment.
```

Comment:
This depends on whether we are the recipient's LSP or the sender's LSP. The current state of our code is weird, because we only have the beginning of the recipient's LSP code, and it's incomplete...I'm tempted to just remove this async payment part for now and reintroduce it later when we have a clearer spec.

Reply:
I removed the async payment part in the latest rebase.